syntax = "proto3";

package auth;

import "google/protobuf/empty.proto";

service Auth {
    rpc RegisterUser (AuthRegisterRequest) returns (AuthRegisterResponse);
    rpc LoginUser (AuthLoginRequest) returns (AuthLoginResponse);
    rpc RefreshAccessToken (RefreshAccessTokenRequest) returns (RefreshAccessTokenResponse);
    rpc CurrentUser (CurrentUserRequest) returns (CurrentUserResponse);
    rpc HealthCheck (google.protobuf.Empty) returns (HealthCheckResponse);
    rpc CreateUserByAdmin (CreateUserByAdminRequest) returns (CreateUserByAdminResponse);
    rpc UpdateUserByAdmin (UpdateUserByAdminRequest) returns (UpdateUserByAdminResponse);
    rpc DeleteUserByAdmin (DeleteUserByAdminRequest) returns (DeleteUserByAdminResponse);
    rpc GetAllUsersByAdmin (GetAllUsersByAdminRequest) returns (GetAllUsersByAdminResponse);
}

message AuthRegisterRequest {
    string username = 1;
    string email = 2;
    string password = 3;
}

message AuthRegisterResponse {
    string id = 1;
    string username = 2;
    string email = 3;
}

message AuthLoginRequest {
    string username = 1;
    string password = 2;
}

message AuthLoginResponse {
    string access_token = 1;
    string refresh_token = 2;
}

message RefreshAccessTokenRequest {
    map<string, string> payload = 1;
}

message RefreshAccessTokenResponse {
    string access_token = 1;
}

message CurrentUserRequest {
    map<string, string> payload = 1;
}

message User {
    string id = 1;
    string username = 2;
    string email = 3;
    bool is_active = 4;
    bool is_superuser = 5;
}

message CurrentUserResponse {
    User user = 1;
}

message HealthCheckResponse {
    string status = 1;
    map<string, string> checks = 2;
}

message CreateUserByAdminRequest {
    string access_token = 1;
    string username = 2;
    string email = 3;
    string password = 4;
    bool is_active = 5;
    bool is_superuser = 6;
}

message CreateUserByAdminResponse {
    string id = 1;
    string username = 2;
    string email = 3;
    bool is_active = 4;
    bool is_superuser = 5;
}

message UpdateUserByAdminRequest {
    string access_token = 1;
    string id = 2;
    optional string username = 3;
    optional string email = 4;
    optional string password = 5;
    optional bool is_active = 6;
    optional bool is_superuser = 7;
}

message UpdateUserByAdminResponse {
    string id = 1;
    string username = 2;
    string email = 3;
    bool is_active = 4;
    bool is_superuser = 5;
}

message DeleteUserByAdminRequest {
    string access_token = 1;
    string id = 2;
}

message DeleteUserByAdminResponse {
    string id = 1;
    bool is_deleted = 2;
}

message GetAllUsersByAdminRequest {
    string access_token = 1;
}

message GetAllUsersByAdminResponse {
    repeated User users = 1;
}